cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,483 @@
+    });
+
+    std::strncpy(name, name_template, kFileNameSize);
+
+#ifdef __APPLE__
+    int fd = mkstemp(name);
+    if(fd == kInvalidHandle)
+        return Status::IOError(strerror(errno));
+    if(fcntl(fd, F_NOCACHE, 1) == -1)
+        return Status::IOError(strerror(errno));
+#else
+    // If we failed, it's possible the temp directory didn't like O_DIRECT,
+    // so we try again without O_DIRECT, and if it still doesn't work then
+    // give up.
+    int fd = mkostemp(name, O_DIRECT);
+    if(fd == kInvalidHandle)
+    {
+        std::strncpy(name, name_template, kFileNameSize);
+        fd = mkstemp(name);
+        if(fd == kInvalidHandle)
+            return Status::IOError(strerror(errno));
+    }
+#endif
+
+    if(unlink(name) != 0)
+        return Status::IOError(strerror(errno));
+    return fd;
+}
+
+static Status CloseTemporaryFile(FileHandle handle)
+{
+    if(close(handle) == -1)
+        return Status::IOError(strerror(errno));
+    return Status::OK();
+}
+
+static Status WriteBatch_PlatformSpecific(FileHandle handle, SpillFile::BatchInfo &info)
+{
+    if(pwritev(handle, info.ios.data(), static_cast<int>(info.ios.size()), info.start) == -1)
+        return Status::IOError("Failed to spill!");
+
+    // Release all references to the buffers, freeing them. 
+    for(ArrayInfo &arr : info.arrays)
+        for(int i = 0; i < 3; i++)
+            if(arr.bufs[i])
+                arr.bufs[i].reset();
+    return Status::OK();
+}
+
+static Result<std::shared_ptr<ArrayData>> ReconstructArray(
+    size_t &idx,
+    SpillFile::BatchInfo &info)
+{
+    ArrayInfo &arr = info.arrays[idx++];
+    std::shared_ptr<ArrayData> data = std::make_shared<ArrayData>();
+    data->type = std::move(arr.type);
+    data->length = arr.length;
+    data->null_count = arr.null_count;
+    data->dictionary = std::move(arr.dictionary);
+    data->buffers.resize(3);
+    for(int i = 0; i < 3; i++)
+        if(arr.sizes[i])
+            data->buffers[i] = std::move(arr.bufs[i]);
+
+    data->child_data.resize(arr.num_children);
+    for(int i = 0; i < arr.num_children; i++)
+    {
+        ARROW_ASSIGN_OR_RAISE(data->child_data[i], ReconstructArray(idx, info));
+    }
+    return data;
+}
+
+static Result<ExecBatch> ReadBatch_PlatformSpecific(
+    FileHandle handle,
+    SpillFile::BatchInfo &info)
+{
+    std::vector<Datum> batch;
+    // ReconstructArray increments i
+    for(size_t i = 0; i < info.arrays.size();)
+    {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayData> ad, ReconstructArray(i, info));
+        batch.emplace_back(std::move(ad));
+    }
+
+    if(preadv(handle, info.ios.data(), static_cast<int>(info.ios.size()), info.start) == -1)
+        return Status::IOError(std::string("Failed to read back spilled data: ") + std::strerror(errno));
+
+    return ExecBatch::Make(std::move(batch));
+}
+#endif
+
+    SpillFile::~SpillFile()
+    {
+        for(BatchInfo *b : batches_)
```

Comment:
Shouldn't we close the file handle here?
Delete the file if needed.

Reply:
File gets deleted automatically when closed (it is `unlink`ed)